File: backend/app/models/product.py
from datetime import datetime, timezone
from typing import List, Optional, Annotated
from pydantic import BaseModel, Field, ConfigDict, BeforeValidator

OidStr = Annotated[str, BeforeValidator(lambda v: str(v))]

class Alias(BaseModel):
    text: str
    normalized: str
    language: str = "en"

class Conversion(BaseModel):
    from_unit: str
    to_unit: str
    factor: float

class Product(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    id: OidStr = Field(alias="_id")
    shop_id: str
    display_name: str
    normalized_name: str
    aliases: List[Alias] = []
    category: Optional[str] = None
    base_unit: str
    allowed_units: List[str] = []
    conversions: List[Conversion] = []
    reorder_threshold: float = 0.0
    active: bool = True
    created_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))
    updated_at: datetime = Field(default_factory=lambda: datetime.now(timezone.utc))

File: frontend/src/types/product.ts
// Backend-aligned types. API shapes:
//   GET /products/ -> { products: [{ id, display_name, category, base_unit,
//     reorder_threshold, quantity, status, ... }], total }
//   GET /inventory/transactions -> { transactions: [{ id, product_id,
//     product_name, operation: STOCK_IN|STOCK_OUT|ADJUSTMENT, quantity, unit,
//     source, created_at }], total }

export interface Product {
  id: string;
  display_name: string;
  name?: string;
  category: string;
  base_unit: string;
  allowed_units?: string[];
  reorder_threshold: number;
  quantity?: number;
  status?: string;
  aliases?: ProductAlias[];
  conversions?: UnitConversion[];
}

export interface ProductAlias {
  text: string;
  normalized?: string;
  language?: string;
}

export interface UnitConversion {
  from_unit: string;
  to_unit: string;
  factor: number;
}

export interface ProductFormData {
  display_name: string;
  category: string;
  base_unit: string;
  reorder_threshold: number;
  aliases: string[];
}
